Add support for `proxy_view` and complementary infrastructure

mingxwa opened the pull request:
**Changes**

- Added class template `observer_facade<F>` for observing a `proxy<F>` via a raw pointer.
- Added alias template `proxy_view<F> = proxy<observer_facade<F>>`.
- Added alias template `add_view<F>` in `basic_facade_builder` to allow implicit conversion from `proxy<F>` to `proxy_view<F>` or `proxy_view<const F>`.
- Added unit tests accordingly.

Shuenhoy reviewed (commented):
Thanks for the implementation!

In general, this is what I expected.

I think there can be some additional tests on the behaviors about const:
* const view from const proxy

```cpp
const pro::proxy<TestFacade> p1 = ...;
pro::proxy_view<const TestFacade> p2 = p1;
```

* const view from non const view

```cpp
pro::proxy_view<TestFacade> p1 = ...;
pro::proxy_view<const TestFacade> p2 = p1;
```

* const view from const raw
```
const int a = 0;
pro::proxy_view<const TestFacade> p = &a;
```

A small question: what is the consideration that the view needs to be explicitly added for each facade?  Reducing compilation time and binary size? Avoiding misuses?

mingxwa commented:
@Shuenhoy Thank you for the comments. I added another 2 cases to cover the independent use of `proxy_view` and added a bunch of `std::as_const()` calls to force the construction of a `proxy_view<const F>` from a `const proxy<F>`.

However, I am hesitant to implement conversion from a non-const view to a const view. Consider the following case:

```cpp
#include <cstdio>

#include "proxy.h"

PRO_DEF_MEM_DISPATCH(MemFoo, Foo);

struct FooFacade : pro::facade_builder
    ::add_convention<MemFoo, void(), void() const>
    ::add_view<FooFacade>
    ::add_view<const FooFacade>
    ::build {};

struct MyFoo {
  void Foo() {}
};

int main() {
  MyFoo foo;
  pro::proxy<FooFacade> p1 = &foo;  // OK
  pro::proxy_view<FooFacade> p2 = p1;  // OK, equivalent to p2 = &foo
  pro::proxy_view<const FooFacade> p3 = std::as_const(p1);  // OK, equivalent to p3 = &foo

  // If p3 = p2 is defined to be equivalent to the following line of code, it won't compile.
  // Therefore, I did not define the conversion from proxy_view<F> proxy_view<const F>.
  // p3 = &std::as_const(foo);
}
```

Requiring explicit `add_view` follows the [zero-overhead principle](https://en.cppreference.com/w/cpp/language/Zero-overhead_principle) since allowing conversion from `proxy` to `proxy_view` adds additional metadata to the vtable, which further increases binary size.

Shuenhoy commented:
Thanks for the clarification.

>// If p3 = p2 is defined to be equivalent to the following line of code, it won't compile.
  // Therefore, I prefer not to define the conversion from proxy_view<F> proxy_view<const F>.
  // p3 = &std::as_const(foo);

My understanding is that the root issue stems from that `proxy<const Facade>` is not a thing, which would be something that behaves like ignoring the non-const dispatches (therefore accepting a const raw pointer).  But that will probably be out of scope of the "view" topic.  So I think this pr can skip const view from non-const view conversion for now, and maybe implement it if you would want to add something like `proxy<const Facade>` in the future.

tian-lt commented:
Suggest adding benchmark cases to see the perf of `proxy_view`

mingxwa commented:
@tian-lt I added benchmarks for `proxy_view` vs virtual functions. The result matches my expectations. Compared to `proxy`, `proxy_view` is a little bit slower when the underlying type is small (because it needs to dereference the pointer), and faster when the underlying type is big. Compared to virtual functions, `proxy_view` is still generally faster. See [the generated report](https://github.com/microsoft/proxy/actions/runs/12350485367?pr=218) from the CI build.

tian-lt commented on tests/proxy_view_tests.cpp:
```diff
@@ -0,0 +1,174 @@
+  *p2 += 3;
+  ASSERT_EQ(ToString(*p1), "126");
+  p1.reset();
+  ASSERT_EQ(a, 126);
```
check `p2` as well here.

tian-lt commented on tests/proxy_view_tests.cpp:
```diff
@@ -0,0 +1,174 @@
+
+struct TestFacade : pro::facade_builder
+    ::add_convention<pro::operator_dispatch<"+=">, void(int)>
+    ::add_convention<utils::spec::FreeToString, std::string() const>
```
add a non-const version for `FreeToString`?

mingxwa replied:
Done
